src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteExpressionBuilder.cs

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal class CallSiteExpressionBuilder: CallSiteVisitor<ParameterExpression, Expression>
+    {
+        private static readonly MethodInfo CaptureDisposableMethodInfo = GetMethodInfo<Func<ServiceProvider, object, object>>((a, b) => a.CaptureDisposable(b));
+        private static readonly MethodInfo TryGetValueMethodInfo = GetMethodInfo<Func<IDictionary<object, object>, object, object, bool>>((a, b, c) => a.TryGetValue(b, out c));
+        private static readonly MethodInfo AddMethodInfo = GetMethodInfo<Action<IDictionary<object, object>, object, object>>((a, b, c) => a.Add(b, c));
+        private static readonly MethodInfo MonitorEnterMethodInfo = GetMethodInfo<Action<object, bool>>((lockObj, lockTaken) => Monitor.Enter(lockObj, ref lockTaken));
+        private static readonly MethodInfo MonitorExitMethodInfo = GetMethodInfo<Action<object>>(lockObj => Monitor.Exit(lockObj));
+
+        private readonly ParameterExpression _providerParameter = Expression.Parameter(typeof(ServiceProvider));
+
+        private readonly List<Expression> _initializations = new List<Expression>();
+        private readonly List<ParameterExpression> _variables = new List<ParameterExpression>();
+        private readonly List<Expression> _locks = new List<Expression>();
+
+        private readonly Dictionary<Expression, ParameterExpression> _resolvedServices = new Dictionary<Expression, ParameterExpression>();
+        private readonly Dictionary<Expression, LambdaExpression> _captureDisposable = new Dictionary<Expression, LambdaExpression>();
+
+        private ParameterExpression _rootVariable;
+
+        public Expression<Func<ServiceProvider, object>> Build(IServiceCallSite callSite)
+        {
+            var serviceExpression = VisitCallSite(callSite, _providerParameter);
+
+            List<Expression> body = new List<Expression>();
+            body.AddRange(_initializations);
+            foreach (var expression in _locks)
```

Comment:
Since `CaptureDisposable` doesn't ever call back into a compiled expression, I don't think it's too important to translate it into an expression tree. I was more suggesting a `LocklessCaptureDisposable` that could be called if a lock is already taken at the top level (which it likely would be for scoped/singleton services. Of course, it only really matters if re-entering the lock is costly, which even at this level it probably not. That's why I phrased it as "if we're doing this anyway for other things..."